### DISCUSSION TITLE
JavaScript closure and React Context

### DISCUSSION THREAD

**Discussion by badboy731** [webpack/webpack#16946] [Q&A] [answered]
How JavaScript closure is used in global state management using React Context?

**Comment by smartcoder0305** [comment, marked as answer]
In React, context allows data to be passed down the component tree without having to pass props manually at every level. It's a way to provide a global state accessible to all components in the application.

JavaScript closures can be used in global state management with React Context by creating a function that returns a new instance of the context. The context's state can be modified using a reducer function. The reducer function can access the previous state and the action dispatched to update the state.

Here's an example:

```javascript
import React, { createContext, useReducer } from 'react';

// Create a new context instance and export it
export const MyContext = createContext();

// Create a reducer function to update the context state
const myReducer = (state, action) => {
  switch (action.type) {
    case 'INCREMENT':
      return { count: state.count + 1 };
    case 'DECREMENT':
      return { count: state.count - 1 };
    default:
      return state;
  }
};

// Create a custom hook to access the context and the dispatch function
export const useMyContext = () => {
  const [state, dispatch] = useReducer(myReducer, { count: 0 });

  // Wrap the `dispatch` function in a JavaScript closure
  const increment = () => dispatch({ type: 'INCREMENT' });
  const decrement = () => dispatch({ type: 'DECREMENT' });

  // Return the current state and `increment`/`decrement` functions
  return [state, { increment, decrement }];
};

```

In this code, we create a new context named `MyContext` and a reducer function `myReducer` to update the state of the context. We then create a custom hook named `useMyContext` that returns the current state and the `increment` and `decrement` functions wrapped in a closure. This allows the context state to be modified by calling the `increment` and `decrement` functions from any component that consumes this context.

To use this context in a component, we need to wrap it in a `Provider` component and pass it the state value:

```javascript
import React from 'react';
import { MyContext, useMyContext } from './my-context';

const Counter = () => {
  const [state, { increment, decrement }] = useMyContext();

  return (
    <div>
      <h2>Count: {state.count}</h2>
      <button onClick={increment}>Increment</button>
      <button onClick={decrement}>Decrement</button>
    </div>
  );
};

const App = () => (
  // Wrap the `MyContext.Provider` around the `Counter`
  <MyContext.Provider>
    <Counter />
  </MyContext.Provider>
);

export default App;

```

In this code, we import the `MyContext` and `useMyContext` we created earlier. We use the `useMyContext` hook to access the context state and functions. The `Counter` component renders the current count and two buttons to increment and decrement it. Finally, we wrap the `Counter` component with the `MyContext.Provider` component.

Now, any component that needs access to the global state can use the `useMyContext` hook to modify and retrieve the state using the `increment` and `decrement` functions.

**Reply by badboy731** [reply]
Thanks. 🙏

**Reply by smartcoder0305** [reply]
You are welcome.
